compare-flatcar-with-gentoo

```diff
@@ -0,0 +1,742 @@
+
+    if [[ -z "${num_ref}" ]]; then
+        num_ref='0'
+    fi
+}
+
+get_plain_version() {
+    local ver="${1}"
+    local arr_name="${2}"
+
+    # strip everything after first underline or dash, whichever comes
+    # first, so for strings like 1.2.3_alpha or 1.2.3-r2 we get 1.2.3
+    ver="${ver%%[-_]*}"
+    IFS='.' read -ra "${arr_name}" <<< "${ver}"
+
+    local i
+    local -n arr_ref="${arr_name}"
+
+    for i in $(seq 0 "$((${#arr_ref[@]}-1))"); do
+        strip_leading_zeros "${arr_name}[${i}]"
+    done
+}
+
+# last version field may have a letter
+split_last_version_field() {
+    local field="${1}"
+    local num_name="${2}"
+    local word_name="${3}"
+
+    local -n num_ref="${num_name}"
+    local -n word_ref="${word_name}"
+
+    num_ref="${field%%*([a-z])}"
+    word_ref="${field##*([0-9])}"
+
+    strip_leading_zeros "${num_name}"
+}
+
+get_suffixes() {
+    local ver="${1}"
+    local arr_name="${2}"
+    local ver_tmp
+
+    if [[ ! "${ver}" =~ _ ]]; then
+        local -n arr_ref="${arr_name}"
+        arr_ref=()
+        return
+    fi
+    ver_tmp="${ver#*_}"
+    ver="${ver_tmp%-*}"
+    IFS='_' read -ra "${arr_name}" <<< "${ver}"
+}
+
+split_suffix() {
+    local suffix="${1}"
+    local word_name="${2}"
+    local num_name="${3}"
+
+    local -n word_ref="${word_name}"
+    local -n num_ref="${num_name}"
+
+    word_ref="${suffix%%*([0-9])}"
+    num_ref="${suffix##*([a-z])}"
+
+    strip_leading_zeros "${num_name}"
+}
+
+get_revision() {
+    local version="${1}"
+    local rev_name="${2}"
+
+    local -n rev_ref="${rev_name}"
+
+    if [[ ! "${version}" =~ '-' ]]; then
+        rev_ref='0'
+        return
+    fi
+    rev_ref="${version##*-r}"
+
+    strip_leading_zeros "${rev_name}"
+}
+
+declare -A SUFFIXES=(
+    ['alpha']='0'
+    ['beta']='1'
+    ['pre']=2
+    ['rc']=3
+    # empty string can't be a key
+    # ['']=4
+    ['p']=5
+)
+
+get_suffix_word_index() {
+    local word="${1}"
+    local index_ref="${2}"
+
+    local -n index="${index_ref}"
+
+    # we can't put an empty key into the hash map
+    if [[ -z "${word}" ]]; then
+        index=4
+        return
+    fi
+    # assign -1 to invalid suffix
+    index="${SUFFIXES[${word}]--1}"
+}
+
+gentoo_vercmp() {
+    local a_ver="${1}"
+    local b_ver="${2}"
+    local res_name="${3}"
+
+    local -n res_ref="${res_name}"
+
+    local -a a_nums
+    local -a b_nums
+
+    get_plain_version "${a_ver}" a_nums
+    get_plain_version "${b_ver}" b_nums
+
+    if [[ "${#a_nums[@]}" -ne "${#b_nums[@]}" ]]; then
+        if [[ "${#a_nums[@]}" -gt "${#b_nums[@]}" ]]; then
+            append_with "${#a_nums[@]}" 0 b_nums
+        else
+            append_with "${#b_nums[@]}" 0 a_nums
+        fi
+    fi
+
+    local i
+    local a
+    local b
+
+    # last field needs to be checked separately
+    for i in $(seq 0 $(("${#a_nums[@]}" - 2))); do
+        a="${a_nums[${i}]}"
+        b="${b_nums[${i}]}"
+
+        if [[ "${a}" -gt "${b}" ]]; then
```

Comment:
I get this error `line 274: [[: 18b: value too great for base (error token is "18b")`

Reply:
Interesting. I did `find <LOCATION> -name '*18b*'` in coreos-overlay, portage-stable and gentoo and got no results (outside `.git` directory, that is). Could you tell me how did you invoke the script and what were the branches (or hashes) of portage-stable, coreos-overlay and gentoo?

Reply:
Basically, this error means that there was a package that had a version like `1.18b.5`, which I think should be invalid, since letters can be a part only of the last version segment (like `1.18.5b` or `1.18b`).

Reply:
Portage-stable is on `7fa8bf336ef784e58689f4f005bd84ea703358ea`, coreos-overlay on `a185ee62ef9fdae7b39f90c3c7951eb9df27a9ea`, gentoo on `d717d5855f276e71c0089cdd2416405e772de9ea`

Reply:
Maybe this here:
```
find gentoo -name '*18b*'
gentoo/sys-apps/lshw/files/lshw-02.18b-gentoo.patch
gentoo/sys-apps/lshw/files/lshw-02.18b-gettext-array.patch
gentoo/sys-apps/lshw/files/lshw-02.18b-sgx.patch
gentoo/sys-apps/lshw/lshw-02.18b-r1.ebuild
gentoo/sys-apps/lshw/lshw-02.18b.ebuild
```

Reply:
Alright, got the same error, thanks. Will debug. But seriously - you have gentoo from `Tue Jul 21 20:46:17 2020 +1000`? :)

Reply:
Fixed.